lib/std/rand.zig

```diff
@@ -247,18 +249,51 @@ pub const Random = struct {
 
     /// Return a floating point value evenly distributed in the range [0, 1).
     pub fn float(r: Random, comptime T: type) T {
-        // Generate a uniform value between [1, 2) and scale down to [0, 1).
-        // Note: The lowest mantissa bit is always set to 0 so we only use half the available range.
+        // Generate a uniformly random value between for the mantissa.
+        // Then generate an exponentially biased random value for the exponent.
+        // Over the previous method, this has the advantage of being able to
+        // represent every possible value in the available range.
         switch (T) {
             f32 => {
-                const s = r.int(u32);
-                const repr = (0x7f << 23) | (s >> 9);
-                return @bitCast(f32, repr) - 1.0;
+                // Use 23 random bits for the mantissa, and the rest for the exponent.
+                // If all 41 bits are zero, generate additional random bits, until a
+                // set bit is found, or 126 bits have been generated.
+                const rand = r.int(u64);
+                var rand_lz = @clz(u64, rand | 0x7FFFFF);
+                if (rand_lz == 41) {
+                    rand_lz += @clz(u64, r.int(u64));
+                    if (rand_lz == 41 + 64) {
+                        // It is astronomically unlikely to reach this point.
```

Comment:
Then wouldn't it be good to use `@setCold` or something here? It probably doesn't work here though because this is not in a function. Maybe if you extract this part into an `inline` function and use `@setCold(true)` there?

Reply:
`@setCold` is function specific and I don't know how well `cold` attributes work on inline functions, better to just add a comment waiting for #5177 or #489 to be implemented.

Reply:
Ah, I didn't know about those. Yes, those seem perfect for this situation.
```suggestion
                        // TODO: when #5177 or #489 is implemented
                        // tell the compiler it is astronomically unlikely to reach this point.
```